osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs

```diff
@@ -1,19 +1,528 @@
+
+                    var newPosition = Vector2.Add(lastPos, relativePos);
+
+                    obj.Position = newPosition;
+
+                    clampToPlayfield(obj);
+
+                    tryCount++;
+                    if (tryCount % 10 == 0) distance *= 0.9f;
+                } while (distance >= obj.Radius * 2 && isOverlappingWithRecent(hitObjects, i));
+
+                if (obj.LastInCombo)
+                    direction = MathHelper.TwoPi * nextSingle();
+                else
+                    direction += distance / distance_cap * (nextSingle() * MathHelper.TwoPi - MathHelper.Pi);
+            }
+        }
+
+        #endregion
+
+        #region Metronome (IApplicableToDrawableRuleset)
+
+        public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
+        {
+            drawableRuleset.Overlays.Add(new TargetBeatContainer());
+        }
+
+        public class TargetBeatContainer : BeatSyncedContainer
+        {
+            private PausableSkinnableSound sample;
+
+            public TargetBeatContainer()
+            {
+                Divisor = 1;
+            }
+
+            protected override void OnNewBeat(int beatIndex, TimingControlPoint timingPoint, EffectControlPoint effectPoint, ChannelAmplitudes amplitudes)
+            {
+                base.OnNewBeat(beatIndex, timingPoint, effectPoint, amplitudes);
+
+                if (!IsBeatSyncedWithTrack) return;
+
+                sample?.Play();
+            }
+
+            [BackgroundDependencyLoader]
+            private void load()
+            {
+                InternalChildren = new Drawable[]
+                {
+                    sample = new PausableSkinnableSound(new SampleInfo("Gameplay/nightcore-hat")) // todo: use another sample
+                };
+            }
+        }
+
+        #endregion
+
+        #region Helper Subroutines
+
+        /// <summary>
+        /// Get samples (if any) for a specific point in time.
+        /// </summary>
+        /// <remarks>
+        /// Samples will be returned if a hit circle or a slider node exists at that point of time.
+        /// </remarks>
+        /// <param name="hitObjects">The list of hit objects in a beatmap, ordered by StartTime</param>
+        /// <param name="time">The point in time to get samples for</param>
+        /// <returns>Hit samples</returns>
+        private IList<HitSampleInfo> getSamplesAtTime(IEnumerable<OsuHitObject> hitObjects, double time)
+        {
+            var sampleObj = hitObjects.FirstOrDefault(x =>
+            {
+                if (Precision.AlmostEquals(time, x.StartTime)) return true;
+
+                if (!(x is Slider s))
+                    return false;
+                if (!Precision.AlmostBigger(time, s.StartTime)
+                    || !Precision.AlmostBigger(s.EndTime, time))
+                    return false;
+
+                return Precision.AlmostEquals((time - s.StartTime) % s.SpanDuration, 0);
```

Comment:
The modulo op on floats is dangerous, especially if we're talking about precision. With fractions that are fundamentally unrepresentable using floating point arithmetic you can get unexpected results. Example:

```
12.1 % 0.1 = 0.09999999999999898
 0.2 % 0.1 = 0
-0.1 % 0.1 = -0
-0.3 % 0.1 = -0.09999999999999998
```

So doing things this way may not prove to be a good idea. You may want to consider doing division here and checking the fractional part of the dividend or something of the sort. Compare mania's `DistanceObjectPatternGenerator.nodeSamplesAt()`.